js/util/struct_array.js

```diff
@@ -0,0 +1,334 @@
+ * @private
+ */
+function serializeStructArrayType() {
+    return {
+        members: this.prototype.StructType.prototype.members,
+        alignment: this.prototype.StructType.prototype.alignment,
+        BYTES_PER_ELEMENT: this.prototype.BYTES_PER_ELEMENT
+    };
+}
+
+
+function align(offset, size) {
+    return Math.ceil(offset / size) * size;
+}
+
+function sizeOf(type) {
+    return viewTypes[type].BYTES_PER_ELEMENT;
+}
+
+function getArrayViewName(type) {
+    return type.toLowerCase();
+}
+
+
+/*
+ * > I saw major perf gains by shortening the source of these generated methods (i.e. renaming
+ * > elementIndex to i) (likely due to v8 inlining heuristics).
+ * - lucaswoj
+ */
+function createEmplaceBack(members, BYTES_PER_ELEMENT) {
+    var usedTypeSizes = [];
+    var argNames = [];
+    var body = '' +
+    'var i = this.length;\n' +
+    'this.length++;\n' +
+    'if (this.length > this.capacity) this._resize(this.length);\n';
+
+    for (var m = 0; m < members.length; m++) {
+        var member = members[m];
+        var size = sizeOf(member.type);
+
+        if (usedTypeSizes.indexOf(size) < 0) {
+            usedTypeSizes.push(size);
+            body += 'var o' + size.toFixed(0) + ' = i * ' + (BYTES_PER_ELEMENT / size).toFixed(0) + ';\n';
+        }
+
+        for (var c = 0; c < member.components; c++) {
+            var argName = 'v' + argNames.length;
+            var index = 'o' + size.toFixed(0) + ' + ' + (member.offset / size + c).toFixed(0);
+            body += 'this.' + getArrayViewName(member.type) + '[' + index + '] = ' + argName + ';\n';
+            argNames.push(argName);
+        }
+    }
+
+    body += 'return i;';
+
+    return new Function(argNames, body);
+}
+
+function createMemberComponentString(member, component) {
+    var elementOffset = 'this._pos' + sizeOf(member.type).toFixed(0);
+    var componentOffset = (member.offset / sizeOf(member.type) + component).toFixed(0);
+    var index = elementOffset + ' + ' + componentOffset;
+    return 'this._structArray.' + getArrayViewName(member.type) + '[' + index + ']';
+
+}
+
+function createGetter(member, c) {
+    return new Function([], 'return ' + createMemberComponentString(member, c) + ';');
+}
+
+function createSetter(member, c) {
+    return new Function(['x'], createMemberComponentString(member, c) + ' = x;');
+}
+
+/**
+ * @class Struct
+ * @param {StructArray} structArray The StructArray the struct is stored in
+ * @param {number} index The index of the struct in the StructArray.
+ * @private
+ */
+function Struct(structArray, index) {
+    this._structArray = structArray;
+    this._setIndex(index);
+}
+
+/**
+ * Make this Struct object point to a different instance in the same array.
+ * It can be cheaper to use .setIndex to re-use an existing Struct than to
+ * create a new one.
+ * @param {number} index The index of the struct in the StructArray;
+ * @private
+ */
+Struct.prototype._setIndex = function(index) {
```

Comment:
I understand now. Renaming to 'getReuse' sounds logical. 

> On Mar 18, 2016, at 2:34 PM, Ansis Brammanis <notifications@github.com> wrote:
> 
> In js/util/struct_array.js:
> 
> > + * @param {number} index The index of the struct in the StructArray.
> > + * @private
> > + */
> > +function Struct(structArray, index) {
> > +    this._structArray = structArray;
> > +    this._setIndex(index);
> > +}
> > +
> > +/**
> > + * Make this Struct object point to a different instance in the same array.
> > + * It can be cheaper to use .setIndex to re-use an existing Struct than to
> > + * create a new one.
> > + * @param {number} index The index of the struct in the StructArray;
> > + * @private
> > + */
> > +Struct.prototype._setIndex = function(index) {
> Your comment makes me think I'm misunderstanding this method. My understanding is that it changes the position at which new elements will be inserted to allow an ArrayBuffer to be shared between StructArrays. Is that right?
> 
> Nope. It's an odd method and its not documented clearly though
> It's an alternative to array.get that doesn't create a new Struct object.
> 
> getReuse smells like premature optimization to me. I've found v8 to be faster declaring new things than mutating existing things in some cases, probably because it can inline them or stack-allocate them instead of heap-allocate them.
> 
> It was added after profiling showed that it was significantly faster in CollisionTile. I'll double check that it still is.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>